### ISSUE TITLE
ENH: signal.ShortTimeFFT: add stream processing capability

### ISSUE THREAD

**Issue by F-M-F-B** [scipy/scipy#22206]
### Is your feature request related to a problem? Please describe.

This functionality is needed to process incoming signals (e.g. audio, sensor measurements) on the fly.

### Describe the solution you'd like.

I'd like to have the option to use an internal buffer in SFT to avoid having to handle samples outside the SFT. stft should add incoming data to buffer and calculate the stft if the buffer contains enough samples.

```py
SFT = ShortTimeFFT(w, hop=10, fs=1/T_x, mfft=200, scale_to='magnitude', buffered=True)
input_device = DataSource()
output = DataSink()

while True:
    x = input_device.read_chunk()
    spectra = SFT.stft(x) # can be empty array up to multiple spectra
    output.write_chunk(spectra)
```

The API of the buffered mode is just an example, there might be better alternatives.

### Describe alternatives you've considered.

The alternative is creating a buffer manually and processing the buffer chunkwise. If window overlapping is required, it is probably easier to use classic fft function instead of ShortTimeFFT.

**Comment by rgommers** [comment]
Thanks for the suggestion @F-M-F-B.
 
> I'd like to have the option to use an internal buffer in SFT to avoid having to handle samples outside the SFT. 

It seems to me like you could make that comment for many functions in `scipy.signal` and `scipy.fft`? It is a feature SciPy doesn't provide anywhere, so it seems odd to me to try and support it in only a single function. Is there a reason that this cannot be done outside of SciPy?

**Comment by F-M-F-B** [comment]
> It seems to me like you could make that comment for many functions in `scipy.signal` and `scipy.fft`? It is a feature SciPy doesn't provide anywhere, so it seems odd to me to try and support it in only a single function. 
In my opinion, the difference between other functions and this one is the statefulness. All functions in fft are "standalone" and do not have an associated class instance. The only other potential candidate would be ZoomFFT but I think there are many more use cases for ShortTimeFFT.
To put it briefly, the statefulness of the ShortTimeFFT makes me wanting that functionality inside SciPy. Implementing this feature for most other functions would add state to them, here the API already requires a state.

> Is there a reason that this cannot be done outside of SciPy?
No, it can easiliy be done outside of SciPy. Although I think the use case of "real time" application is not uncommon, so handling this outside SciPy leads to duplicated code/functionality across projects and users.

**Comment by DietBru** [comment]
When writing the ShortTimeFFT, I pondered quite a bit on how to implement a generic incremental processing (or streaming) functionality. I came to the conclusion that custom data management requirements, like streams, are better handled outside the ShortTimeFFT class.  The reason is that for different applications, there exist contradicting objectives, which call for different implementations, e.g.:
* Minimize latency by calculating each slice as soon there is enough data versus maximizing data throughput by utilizing the multiprocessing capabilities of the FFT backend.  
* Minimize RAM usage by implementing a stream architecture versus keeping all data in RAM (e.g., for a recording audio editor)
* ...

Hand-rolling your own solution boils down to reimplementing this [for loop]:

```python
        S = np.zeros(S_shape, dtype=complex)
        for p_, x_ in enumerate(self._x_slices(x, k_offset, p0, p1, padding)):
            if detr is not None:
                x_ = detr(x_)
            S[..., :, p_] = self._fft_func(x_ * self.win.conj())
```
The [_x_slices] method just provides slices of the input array, as described [here]. 

Unless there is a widely used streaming format/implementation, which would be useful to support, I would vote to close this issue, as "not planned". Does this make sense, @F-M-F-B?

[for loop]: https://github.com/scipy/scipy/blob/31cbb91f22305e0311026f199213e01eac7a8470/scipy/signal/_short_time_fft.py#L845-L849
[_x_slices]: https://github.com/scipy/scipy/blob/31cbb91f22305e0311026f199213e01eac7a8470/scipy/signal/_short_time_fft.py#L712
[here]: https://docs.scipy.org/doc/scipy/tutorial/signal.html#sliding-windows
